Src/FluentAssertions/EventRaisingExtensions.cs

```diff
@@ -13,84 +15,171 @@ namespace FluentAssertions
+                    hasArgumentOfRightType = true;
+                }
+
+                if (typedParameters.Any(parameter => compiledPredicate(parameter)))
+                {
+                    eventsMatchingPredicate.Add(@event);
+                }
             }
 
-            if (eventRecorder.All(recordedEvent => !recordedEvent.Parameters.OfType<T>().Any(parameter => compiledPredicate(parameter))))
+
+            if (!hasArgumentOfRightType)
+            {
+                throw new ArgumentException("No argument of event " + eventRecording.EventName + " is of type <" + typeof(T) + ">.");
+            }
+
+            if (!eventsMatchingPredicate.Any())
             {
                 Execute.Assertion
                     .FailWith("Expected at least one event with arguments matching {0}, but found none.", predicate.Body);
             }
 
-            return eventRecorder;
+            return new FilteredEventRecording(eventRecording, eventsMatchingPredicate);
         }
 
         /// <summary>
-        /// Asserts that at least one occurrence of the event had arguments matching all predicates.
+        /// Asserts that at least one occurrence of the events had arguments matching all predicates. Returns
+        /// only the events that matched that predicate.
         /// </summary>
-        public static IEventRecorder WithArgs<T>(this IEventRecorder eventRecorder, params Expression<Func<T, bool>>[] predicates)
+        public static IEventRecording WithArgs<T>(this IEventRecording eventRecording, params Expression<Func<T, bool>>[] predicates)
         {
             Func<T, bool>[] compiledPredicates = predicates.Select(p => p?.Compile()).ToArray();
 
-            if (!eventRecorder.First().Parameters.OfType<T>().Any())
-            {
-                throw new ArgumentException("No argument of event " + eventRecorder.EventName + " is of type <" + typeof(T) + ">.");
-            }
+            bool hasArgumentOfRightType = false;
+            var eventsMatchingPredicate = new List<OccurredEvent>();
 
-            bool expected = eventRecorder.Any(recordedEvent =>
+            foreach (OccurredEvent @event in eventRecording)
             {
-                T[] parameters = recordedEvent.Parameters.OfType<T>().ToArray();
-                int parametersToCheck = Math.Min(parameters.Length, predicates.Length);
+                var typedParameters = @event.Parameters.OfType<T>().ToArray();
+                if (typedParameters.Any())
+                {
+                    hasArgumentOfRightType = true;
+                }
+
+                int parametersToCheck = Math.Min(typedParameters.Length, predicates.Length);
 
                 bool isMatch = true;
-                for (int i = 0; i < parametersToCheck && isMatch; i++)
+                for (int index = 0; index < parametersToCheck && isMatch; index++)
+                {
+                    isMatch = compiledPredicates[index]?.Invoke(typedParameters[index]) ?? true;
```

Comment:
Just to double check, is it intended that when `compiledPredicates[index]` returns null `isMatch` should be true?

Reply:
Yes, because you can pass `null` for a predicate to ignore that value.